scratchml/models/svc.py

```diff
@@ -0,0 +1,202 @@
+            )
+
+    def _validate_parameters(self):
+        if self.C <= 0:
+            raise ValueError("Regularization parameter C must be positive.")
+
+    def _check_is_fitted(self):
+        if self.weights is None:
+            raise ValueError("Model must be trained before prediction.")
+
+    def fit(self, X, y):
+        """
+        Trains the SVC model using mini-batch gradient descent.
+
+        Parameters
+        ----------
+        X : ndarray, shape (n_samples, n_features)
+            Training data.
+        y : ndarray, shape (n_samples,)
+            Target labels (e.g., 0 or 1 for binary classification).
+        """
+        self.classes_ = np.unique(y)
+        if len(self.classes_) == 2:
+            y_ = np.where(y == self.classes_[0], -1, 1)
+        else:
+            y_ = y
+
+        X = self.scaler.fit_transform(X)
+        n_samples, n_features = X.shape
+        self.weights = np.zeros(n_features)
+
+        no_improvement_count = 0
+        for iteration in range(self.max_iter):
+            lr = self.learning_rate * (self.decay**iteration)
+            avg_update_norm = 0
+            for X_batch, y_batch in split_data_into_batches(
+                X, y_, self.batch_size, shuffle=True
+            ):
+                weight_update = np.zeros_like(self.weights)
+                bias_update = 0
+                for idx in range(X_batch.shape[0]):
+                    instance = X_batch[idx : idx + 1]
+                    margin = y_batch[idx] * (
+                        np.dot(instance.flatten(), self.weights) - self.bias
+                    )
+
+                    if margin < 1:
+                        weight_update += -self.C * y_batch[idx] * instance.flatten()
+                        bias_update += -self.C * y_batch[idx]
+
+                weight_update += self.alpha * l2(self.weights)
+
+                lr_adjusted = (
+                    lr / (1 + np.linalg.norm(weight_update)) if self.adaptive_lr else lr
+                )
+                self.weights -= lr_adjusted * weight_update
+                self.bias -= lr_adjusted * bias_update
+                avg_update_norm += np.linalg.norm(weight_update)
+
+            avg_update_norm /= X.shape[0] / self.batch_size
+            if self.early_stopping and avg_update_norm < self.tol:
+                no_improvement_count += 1
+                if (
+                    no_improvement_count > 10
+                ):  # Terminate if no improvement for 10 iterations
+                    # print(f"Converged after {iteration} iterations.")
+                    break
+            else:
+                no_improvement_count = 0
+
+        #     if iteration % 100 == 0:
+        #         print(f"Iteration {iteration}: Avg norm of batch updates = {avg_update_norm}")
+
+        # print(f"Final weight norm: {np.linalg.norm(self.weights)}")
+        # print(f"Final bias: {self.bias}")
+
+    def predict(self, X):
```

Comment:
Please, add typing here too. Like this:

```python3
def predict(self, X: np.ndarray) -> np.ndarray:
```

Reply:
Please, add a docstring as well. Like this:

```python3
def predict(self, X: np.ndarray) -> np.ndarray:
    """
    Uses the trained model to predict the classes of a given
    data points (also called features).

    Args:
       X (np.ndarray): the features.

    Returns:
       np.ndarray: the predicted classes.
    """
```

You can get the typings and the docstring of that function from other files, such as the Random Forest Classifier code (line 364 in file random_forest.py).